api/service_principals.go

```diff
@@ -15,6 +21,87 @@ type ServicePrincipalClient interface {
 }
 
 type ServicePrincipal struct {
-	ObjectID string
-	AppID    string
+	ID    string
+	AppID string
+}
+
+func (c *MSGraphClient) CreateServicePrincipal(ctx context.Context, appID string, startDate time.Time, endDate time.Time) (string, string, error) {
+	spReq := models.NewServicePrincipal()
+	spReq.SetAppId(&appID)
+
+	sp, err := c.client.ServicePrincipals().Post(ctx, spReq, nil)
+	if err != nil {
+		return "", "", err
+	}
+
+	spID := sp.GetId()
+
+	passwordReq := serviceprincipals.NewItemAddPasswordPostRequestBody()
+	passwordCredential := models.NewPasswordCredential()
+	passwordCredential.SetStartDateTime(&startDate)
+	passwordCredential.SetEndDateTime(&endDate)
+
+	passwordReq.SetPasswordCredential(passwordCredential)
+
+	password, err := c.client.ServicePrincipals().ByServicePrincipalId(*spID).AddPassword().Post(ctx, passwordReq, nil)
+
+	if err != nil {
+		e := c.DeleteServicePrincipal(ctx, *spID, false)
+		merr := multierror.Append(err, e)
+		return "", "", merr.ErrorOrNil()
+	}
+	return *spID, *password.GetSecretText(), nil
+}
+
+func (c *MSGraphClient) DeleteServicePrincipal(ctx context.Context, spObjectID string, permanentlyDelete bool) error {
+	err := c.client.ServicePrincipals().ByServicePrincipalId(spObjectID).Delete(ctx, nil)
+
+	if permanentlyDelete {
+		e := c.client.Directory().DeletedItems().ByDirectoryObjectId(spObjectID).Delete(ctx, nil)
+		merr := multierror.Append(err, e)
+		return merr.ErrorOrNil()
+	}
+
+	return err
+}
+
+func (c *MSGraphClient) ListServicePrincipals(ctx context.Context, spObjectID string) ([]ServicePrincipal, error) {
+	filter := fmt.Sprintf("appId eq '%s'", spObjectID)
+	requestParameters := &serviceprincipals.ServicePrincipalsRequestBuilderGetQueryParameters{
+		Filter: &filter,
+	}
+
+	headers := abstractions.NewRequestHeaders()
+	headers.Add("ConsistencyLevel", "eventual")
+
+	configuration := &serviceprincipals.ServicePrincipalsRequestBuilderGetRequestConfiguration{
+		Headers:         headers,
```

Comment:
I believe we can remove this as well since we are only ever using the `eq` filter.